clients/lsp-ruff-lsp.el

```diff
@@ -0,0 +1,95 @@
+;;; lsp-ruff-lsp.el --- ruff-lsp support             -*- lexical-binding: t; -*-
+
+;; Copyright (C) 2023 Freja Nordsiek
+;;
+;; Author: Freja Nordsiek <fnordsie@posteo.net
+;; Keywords: language tools
+;;
+;; This program is free software; you can redistribute it and/or modify
+;; it under the terms of the GNU General Public License as published by
+;; the Free Software Foundation, either version 3 of the License, or
+;; (at your option) any later version.
+;;
+;; This program is distributed in the hope that it will be useful,
+;; but WITHOUT ANY WARRANTY; without even the implied warranty of
+;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+;; GNU General Public License for more details.
+;;
+;; You should have received a copy of the GNU General Public License
+;; along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+;;; Commentary:
+
+;; ruff-lsp Client for the Python programming language
+
+;;; Code:
+
+(require 'lsp-mode)
+
+(defgroup lsp-ruff-lsp nil
+  "LSP support for Python, using ruff-lsp's Python Language Server."
+  :group 'lsp-mode
+  :link '(url-link "https://github.com/charliermarsh/ruff-lsp"))
+
+(defcustom lsp-ruff-lsp-server-command '("ruff-lsp")
+  "Command to start ruff-lsp."
+  :risky t
+  :type '(repeat string)
+  :group 'lsp-ruff-lsp)
+
+(defcustom lsp-ruff-lsp-ruff-path ["ruff"]
+  "Paths to ruff to try, in order."
+  :risky t
+  :type 'lsp-string-vector
+  :group 'lsp-ruff-lsp)
+
+(defcustom lsp-ruff-lsp-ruff-args []
+  "Arguments, passed to ruff."
+  :risky t
+  :type 'lsp-string-vector
+  :group 'lsp-ruff-lsp)
+
+(defcustom lsp-ruff-lsp-log-level "error"
+  "Tracing level."
+  :type '(choice (const :tag "error" "error")
+                 (const :tag "warning" "warning"))
+  :group 'lsp-ruff-lsp)
+
+(defcustom lsp-ruff-lsp-python-path "python3"
+  "Path to the Python interpreter."
+  :risky t
+  :type 'string
+  :group 'lsp-ruff-lsp)
+
+(defcustom lsp-ruff-lsp-advertize-organize-imports t
+  "Whether to report ability to handle source.organizeImports actions."
+  :type 'boolean
+  :group 'lsp-ruff-lsp)
+
+(defcustom lsp-ruff-lsp-advertize-fix-all t
+  "Whether to report ability to handle source.fixAll actions."
+  :type 'boolean
+  :group 'lsp-ruff-lsp)
+
+
+(lsp-register-client
```

Comment:
shouldn't that be add-on? = t?

Reply:
I don't know. Reading the FAQ, it seems like ``add-on?`` being ``t`` means load in parallel to other LSPs at the same priority? Is compatibility something to worry about? Only a few of the client modules use it:

* lsp-eslint.el
* lsp-emmet.el
* lsp-sorbet.el
* lsp-angular.el
* lsp-camel.el

None of the python ones do at present (though, of course, that doesn't mean that they shouldn't).

So basically, I don't know enough about the option to know.

Reply:
Ruff doesn't provide anything you'd like a language server to have, e.g. code navigation, completion etc., it's just a linter. So running it standalone is usually not what you want, but pairing it with e.g. pylsp is extremely natural -- you'd have two servers running, both providing code actions and diagnostics at the same time (e.g. pylsp would provide mypy diagnostics, while ruff would be linting). It's exactly what we want and need (like I said, it's basically the same situation as with eslint). We haven't seen addon LSPs with Python yet, since pylint and pyflakes and flake8 and all the other linters have been typically used as plugins for pylsp, but ruff is mostly used standalone (but there is a plugin for pylsp too).

Reply:
So that is why the other python LSPs have not done it. Makes sense. Speaking of, the option to enable ruff for pylsp does need to have its customization options added but that is for another day.

Thank you for the explanations.